core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java

```diff
@@ -0,0 +1,299 @@
+  public void testMultipleTxnOperations() {
+    Snapshot snapshot;
+    Transaction txn = table.newTransaction();
+    txn.newOverwrite().addFile(FILE_A).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(1, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    txn = table.newTransaction();
+    Set<DataFile> toAddFiles = new HashSet<>();
+    Set<DataFile> toDeleteFiles = new HashSet<>();
+    toAddFiles.add(FILE_B);
+    toDeleteFiles.add(FILE_A);
+    txn.newRewrite().rewriteFiles(toDeleteFiles, toAddFiles).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(2, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    txn = table.newTransaction();
+    txn.newReplacePartitions().addFile(FILE_C).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(3, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    txn = table.newTransaction();
+    txn.newDelete().deleteFile(FILE_C).commit();
+    txn.commitTransaction();
+
+    Assert.assertEquals(4, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    txn = table.newTransaction();
+    txn.newAppend().appendFile(FILE_C).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(5, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    snapshot = table.currentSnapshot();
+    txn = table.newTransaction();
+    txn.newOverwrite().addFile(FILE_D).deleteFile(FILE_C).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(6, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    txn = table.newTransaction();
+    txn.expireSnapshots().expireOlderThan(snapshot.timestampMillis()).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(6, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+  }
+
+  @Test
+  public void testSequenceNumberForCherryPicking() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // WAP commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .set("wap.id", "123456789")
+        .stageOnly()
+        .commit();
+
+    Assert.assertEquals("the snapshot sequence number should be 1", 1,
+        table.currentSnapshot().sequenceNumber());
+
+    // pick the snapshot that's staged but not committed
+    Snapshot wapSnapshot = readMetadata().snapshots().get(1);
+
+    Assert.assertEquals("the snapshot sequence number should be 2", 2,
+        wapSnapshot.sequenceNumber());
+
+    // table has new commit
+    table.newAppend()
+        .appendFile(FILE_C)
+        .commit();
+
+    Assert.assertEquals("the snapshot sequence number should be 3",
+        3, table.currentSnapshot().sequenceNumber());
+
+    // cherry-pick snapshot
+    table.manageSnapshots().cherrypick(wapSnapshot.snapshotId()).commit();
+
+    Assert.assertEquals("the snapshot sequence number should be 4",
+        4, table.currentSnapshot().sequenceNumber());
```

Comment:
This should also assert that the picked files also have the new sequence number.

Reply:
OK

Reply:
I think this still needs to be updated.

Reply:
The picked file is validated in the next line. Do you mean that？

```validateDataFiles(files(FILE_A, FILE_B, FILE_C), seqs(1, 4, 3));```

Reply:
Yeah, sorry I missed that.